### ISSUE TITLE
/etc/init.d/icinga2 reload && /etc/init.d/icinga2 restart fails on CentOS 6

### ISSUE THREAD

**Issue by hfinucane** [Icinga/icinga2#5312]
I use the icinga2 Chef cookbook, which triggers that internally, and it means that bringing up fresh servers doesn't work reliably. This appears to be a race condition- `/etc/init.d/icinga2 reload && sleep 1 && /etc/init.d/icinga2` works just fine.

I get
```
[2017-05-31 18:01:15 +0000] information/cli: Icinga application loader (version: v2.6.3)
[2017-05-31 18:01:15 +0000] critical/cli: Another instance of Icinga already running with PID 8878
```

in the startup log.

**Comment by dgoetz** [comment]
From an Icinga 2 perspective this restart logic does not make sense. Reload is enough and wont stop checking and alarming, restart only adds the overhead of system initializing the service and will stop Icinga 2's work.

Perhaps `/etc/init.d/icinga2 reload || /etc/init.d/icinga2 restart` would make sense, trying to restart if reload fails, but then start would be better as you would not like to stop if reload fails because of a syntax error.

**Comment by hfinucane** [comment]
That's fair, and possibly I should have said `/etc/init.d/icinga2 reload ;  /etc/init.d/icinga2` instead- the point is that `reload` leaves the service in a busted state for a while which can break automation tools, especially if they're updating the icinga2 package and touching configuration during the same run.

**Comment by dnsmichi** [comment]
There's not much we can do here since the child process takes over in an asynchronous fashion.

What's the idea behind a reload followed by a restart? They're basically the same, just with the bonus that reload won't kill your running process with stop().

**Comment by hfinucane** [comment]
I use the official Chef cookbook, and while [most everything uses reload](https://github.com/Icinga/chef-icinga2/search?utf8=%E2%9C%93&q=%3Areload&type=), [package installation and some other things trigger restarts](https://github.com/Icinga/chef-icinga2/search?utf8=%E2%9C%93&q=%3Arestart&type=). This means that first-runs and package updates can cause Chef-layer crashes, because it doesn't know that `reload` makes `restart` unsafe. I have 'solved' this by vendoring `/usr/lib/icinga2/safe-reload` and making it sleep, but that seems ... not ideal?

I'm hoping that there is a socket or something that `safe-reload` could wait for, and then when it was back, we could 'know' that the child process was done taking over.

**Comment by dnsmichi** [comment]
Unfortunately there is no such socket since the `api` feature for example remains fully optional, and you could run the process just with the checker feature enabled. You could workaround that and always enable the api feature and wait for just that.

I've seen a similar thing with vagrant boxes and elasticsearch - you'll need to wait until the API is fully there to proceed with operations. Solution: An endless curl loop. 

I'm not sure how we may solve the problem with Icinga 2 here. Workarounds do exist, and that's good for the moment.

**Comment by dnsmichi** [comment]
Might be solved with #2287 (PRs welcome). I don't think that such a thing exists for legacy SysVInit and we'll have to deal with it until it gets replaced by Systemd.

Cheers,
Michael
